tests/isa-test.cc

```diff
@@ -0,0 +1,769 @@
+
+		public:
+		SetBoundsTestCase(ptraddr_t aRequestedBase,
+		                  ptraddr_t aRequestedLength,
+		                  ptraddr_t anExpectedBase,
+		                  ptraddr_t anExpectedTop)
+		  : RequestedBase(aRequestedBase),
+		    RequestedLength(aRequestedLength),
+		    ExpectedBase(anExpectedBase),
+		    ExpectedTop(anExpectedTop)
+		{
+		}
+
+		void run_test()
+		{
+			/* Sanity checks on the test case TODO check statically? */
+			TEST(ExpectedBase <= ExpectedTop,
+			     "Invalid test case: exp_base > exp_top");
+			TEST(ExpectedBase <= RequestedBase,
+			     "Invalid test case: exp_base > req_base");
+			TEST(ExpectedTop >= RequestedBase + RequestedLength,
+			     "Invalid test case:  exp_top < req_top");
+
+			/*
+			 * Attempt set bounds with requested bounds. By starting with NULL
+			 * we can check that we get the bounds we expect but can't validate
+			 * that bounds checks / tag clearing are performed correctly. To do
+			 * that we'd need a very large capability to start with.
+			 */
+			Capability<uint8_t> c{NULL};
+			c += RequestedBase; // Technically UB, but seems to work.
+			c.bounds().set_inexact(RequestedLength);
+			debug_log("Validating set bounds result: {}", c);
+			TEST(c.base() == ExpectedBase,
+			     "Unexpected base: expected {} but got {}",
+			     ExpectedBase,
+			     c.base());
+			TEST(c.top() == ExpectedTop,
+			     "Unexpected top: expected {} but got {}",
+			     ExpectedTop,
+			     c.top());
+		}
+	};
+
+	SetBoundsTestCase setBoundsCases[] = {
+	  SetBoundsTestCase(0x100, 0x1a, 0x100, 0x11a),  // easy case
+	  SetBoundsTestCase(0x700, 0x1aa, 0x700, 0x8aa), // easy case
+	  SetBoundsTestCase(0xd01,
+	                    0x3ff,
+	                    0xd00,
+	                    0x1100), // exact top, inexact base, e bump
+	  SetBoundsTestCase(0x9f3, 0x7ff, 0x000009F0, 0x000011F8), // T-B = 0x201
+	  SetBoundsTestCase(0xbeef9793,
+	                    0x3fb,
+	                    0xBEEF9792,
+	                    0xBEEF9B8E), // monotonicity failure regression part i
+	  SetBoundsTestCase(0xbeef9792,
+	                    0x3fc,
+	                    0xBEEF9792,
+	                    0xBEEF9B8E), // monotonicity failure regression part ii
+	  SetBoundsTestCase(0x4fffe1ff,
+	                    0x3fe,
+	                    0x4FFFE1FC,
+	                    0x4FFFE600), // requires T increment
+	  SetBoundsTestCase(0xe7e96c2f,
+	                    0x3ff,
+	                    0xE7E96C2C,
+	                    0xE7E97030), // requires e bump and T increment
+	};
+
+	void test_set_bounds()
+	{
+		int x = 0;
+		for (auto testCase : setBoundsCases)
+		{
+			testCase.run_test();
+		}
+	}
+
+	/**
+	 * Returns a PermissionSet that is the set of permissions that the
+	 * hardware would return if asked to encode PermissionSet `ps` by
+	 * CAndPerms.
+	 *
+	 * The returned PermissionSet will always be a subset of `ps`.
+	 *
+	 * TODO make this a method of PermissionSet?
```

Comment:
Will check with @davidchisnall on this one.

Reply:
Looks generally useful.  Could also wrap it with 

```c++
constexpr bool is_representable() const
{
    return (*this) == to_representable();
}
```

This would be useful in a bunch of static asserts.